x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/inference/results/TextEmbeddingResults.java

```diff
@@ -103,24 +105,60 @@ public Map<String, Object> asMap() {
         return map;
     }
 
-    public record Embedding(List<Float> values) implements Writeable, ToXContentObject {
+    public static class Embedding implements Writeable, ToXContentObject {
         public static final String EMBEDDING = "embedding";
 
+        public static Embedding ofFloats(List<Float> values) {
+            return new Embedding(convertFloatsToEmbeddingValues(values));
+        }
+
+        public static Embedding ofBytes(List<Byte> values) {
+            List<EmbeddingValue> convertedValues = values.stream().map(ByteValue::new).collect(Collectors.toList());
+
+            return new Embedding(convertedValues);
+        }
+
+        private final List<EmbeddingValue> values;
+
+        public Embedding(List<EmbeddingValue> values) {
+            this.values = values;
+        }
+
         public Embedding(StreamInput in) throws IOException {
-            this(in.readCollectionAsImmutableList(StreamInput::readFloat));
+            if (in.getTransportVersion().onOrAfter(TransportVersions.ML_INFERENCE_COHERE_EMBEDDINGS_ADDED)) {
+                values = in.readNamedWriteableCollectionAsList(EmbeddingValue.class);
+            } else {
+                values = convertFloatsToEmbeddingValues(in.readCollectionAsImmutableList(StreamInput::readFloat));
```

Comment:
I don't think we actually want nodes converting bytes into floats when sending to other nodes. I believe we'll prevent that by using the cohere service's minimum support version. So I don't think we'll ever need to execute this code.

Reply:
Yeah good question. Dave and I ran into this with this class as well: https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/inference/action/InferenceAction.java#L215-L224

which should be similar in that a few of those if-blocks shouldn't really happen. We decided for completeness that we'd handle the cases. I'm happy to remove it if we think otherwise though. It just gets complicated with the nodes on separate versions during upgrades. It should be possible but I suppose if we have a bug in the future we could get in a case where it was 🤷‍♂️

Reply:
I think at least we should be adding in an assertion here so that any tests that hit this code path fail

Reply:
With Dave's suggestion I think we can get away with the transport version check all together. It'll add duplicate code though. I'll add a comment on the new class.